### PR TITLE
Updating collectives

### PR DESCRIPTION
**Pull request by tomoleary**
- Added bcast collective
- Support expanded to MultiVector for collectives

### CODE REVIEW

File: hippylib/scheduling/collective.py [hunk 1]
```diff
@@ -116,7 +121,60 @@ def allReduce(self, v, op):
             v.set_local(receive)
             v.apply("")
             
+            return v
+        elif hasattr(v,'nvec'):
+            for i in range(v.nvec()):
+                assert v[i].mpi_comm().Get_size() == 1
```
Comment (uvilla) [inline hippylib/scheduling/collective.py]: I'd suggest a lazier way to handle multivector:

```
        elif hasattr(v,'nvec'):
            for i in range(v.nvec()):
                 self.allReduce(self, v[i], op)
```

File: hippylib/test/ptest_collectives.py [hunk 2]
```diff
@@ -91,8 +120,65 @@ def testdlVector(self):
...
+        # test allReduce
+        mesh = dl.UnitSquareMesh(dl.MPI.comm_self,10, 10)
+        Vh = dl.FunctionSpace(mesh, 'Lagrange', 1)
+
+        x = dl.interpolate(dl.Constant(1.), Vh).vector()
+        ones = np.ones_like(x.get_local())
+        MV = MultiVector(x,10)
+        MV_ref = MultiVector(x,10)
+        for i in range(MV.nvec()):
+            MV[i].set_local(ones)
```
Comment (uvilla) [inline hippylib/test/ptest_collectives.py]: Always call `apply` after `set_local`

File: hippylib/scheduling/comm_utils.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    collective.bcast(root_DG0_v.vector(),root = 0)
+    collective.bcast(root_CG1_v.vector(),root = 0)
+
+    diff_DG0 = DG0_v - root_DG0_v
+    diff_CG1 = CG1_v - root_CG1_v
+
+    assert_allclose( [diff_DG0.norm("l2")], [0.])
+    assert_allclose( [diff_CG1.norm("l2")], [0.])
+
+    if collective_rank == 1:
```
Comment (uvilla) [inline hippylib/scheduling/comm_utils.py]: Remove the print on screen and assert.

Instead make this function return `True` if the partitioning is correct or `False` if it is not.

NOTE: USE MPI_COMM_WORLD to make sure that if one process detect `False` all processes will return `False`

**Comment by tomoleary** [comment]
Remaining issues for the PR

- There are issues with the 'avg' mpi_op in the CollectiveOperator class. Why is this?

- When using split communicators (sample parallelism and mesh partitioning) why does the eigenvalue problem seem to execute on each piece of the mesh as opposed to the whole mesh?

**Comment by tomoleary** [comment]
> Remaining issues for the PR
> 
> * There are issues with the 'avg' mpi_op in the CollectiveOperator class. Why is this?
> * When using split communicators (sample parallelism and mesh partitioning) why does the eigenvalue problem seem to execute on each piece of the mesh as opposed to the whole mesh?

Somehow the mesh gets partitioned when I think I am only doing sample parallelism, so the row and column communicators may have been switched. If this is the case it will be somewhat straightforward to fix.

**Comment by tomoleary** [comment]
Been a while since this PR. Issues with averaged operators still persist for multiple copies of the same mesh. Will look into this more.

**Comment by uvilla** [comment]
Superseded by PR #62 that has been merged in master

**Pull request closed without merging**